gips/0065-subgraph-availability-manager.md

```diff
@@ -0,0 +1,127 @@
+
+# High-Level Description
+
+The new smart contract `SubgraphAvailabilityManager` will be exclusively deployed to the Arbitrum One network and will function as a mediator between five distinct oracle operators and the `RewardsManager` contract. It will feature a vote aggregation system and enforce an execution threshold requiring three votes to alter the state of a subgraph deployment within the `RewardsManager` contract. Additionally, it will incorporate a `voteTimeLimit` timeframe to monitor current votes, after which they expire and become invalid.
+
+It's important to note that this system operates effectively only if a minimum of three oracles are honest and running. The system will be able to report subgraph availability whenever three or more of the oracle instances are running correctly. According to the Charter guidelines, all oracles are expected to operate in compliance with established standards. In the event that any oracles are found not to be performing up to these standards, measures will be taken to replace them, ensuring the system's integrity and continuous effective operation.
+
+The `SubgraphAvailabilityManager` contract will be governed by the council. While the contract itself will not be upgradeable, a new contract can be deployed, and the council can set the `RewardsManager` oracle address to the new deployment. The council will have the authority to add or remove oracles operators and update the `voteTimeLimit` timeframe.
+
+# Detailed Specification
+
+The number of oracle operators will be strictly enforced as five. Upon contract initialization, five valid accounts must be provided, and each oracle must maintain their index position, crucial for passing it as a parameter when casting votes.
+
+Vote aggregation will be separately tracked for positive or negative votes using the following data structures:
+
+```solidity
+mapping(uint256 currentNonce => mapping(bytes32 subgraphDeploymentID => uint256[NUM_ORACLES] timestamps)) public lastDenyVote;
+mapping(uint256 currentNonce => mapping(bytes32 subgraphDeploymentID => uint256[NUM_ORACLES] timestamps)) public lastAllowVote;
+```
+
+The initial mapping key corresponds to a public variable `currentNonce`. This nonce serves as a mechanism to invalidate all votes whenever a configuration is altered—be it updating the oracle operators list or modifying the `voteTimeLimit` parameter. The inner mapping stores the `subgraphDeploymentId` against the timestamp of each oracle's vote, using an array denoting the oracle index.
+
+The `executionThreshold` will be set to three during contract construction. Whenever an oracle casts a vote, the `SubgraphAvailabilityManager` contract will query the corresponding vote aggregation storage to verify if enough votes have been registered. Upon the third vote a transaction to `RewardsManager` is triggered to make the subgraph rewards state change.
```

Comment:
this might be nuanced for the implementer: what happens if the votes are not unanimous? for example, if there's 2 positive and 3 negative votes, should the transaction still be triggered in favor of 3 negative votes without any additional actions?

Reply:
Transaction will still be triggered when reaching the threshold of 3. All oracles should be running the same configs and same code so they should be voting equally. If oracles vote differently, the appropriate action is to investigate the discrepancy and respond accordingly based on the findings of the investigation.